sstables/storage.cc

```diff
@@ -533,6 +540,28 @@ future<> s3_storage::wipe(const sstable& sst, sync_dir) noexcept {
     co_await sys_ks.sstables_registry_delete_entry(_location, sst.generation());
 }
 
+future<> s3_storage::remove_by_registry_entry(utils::UUID uuid, entry_descriptor desc) {
+    auto prefix = format("/{}/{}", _bucket, uuid);
+    std::optional<temporary_buffer<char>> toc;
+    std::vector<sstring> components;
+
+    try {
+        toc = co_await _client->get_object_contiguous(prefix + "/" + sstable_version_constants::get_component_map(desc.version).at(component_type::TOC));
+    } catch (seastar::httpd::unexpected_status_error& e) {
+        if (e.status() != seastar::http::reply::status_type::no_content) {
+            throw;
+        }
+    }
+    if (!toc) {
+        co_return; // missing TOC object is OK
+    }
+
```

Comment:
if toc is missing, perhaps we should try deletion of known sstable components, while forgiving "resource not found" failure? then we avoid leaks which might happen if toc somehow ends up missing for a sstable.

Reply:
That's unfortunate, because opening S3-backed sstable first creates the TOC entry. However, you just spotted a BUG here. If g.c. removes the TOC and crashes, then subsequent g.c. would skip the whole thing altogether, so TOC must be removed last. Will fix.